packages/core/http/core-http-server-internal/src/register_lifecycle_handlers.ts

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type { Env } from '@kbn/config';
+import type { HttpConfig } from './http_config';
+import type { LifecycleRegistrar } from './http_server';
+import {
+  createCustomHeadersPreResponseHandler,
+  createRestrictInternalRoutesPostAuthHandler,
+  createVersionCheckPostAuthHandler,
+  createXsrfPostAuthHandler,
+} from './lifecycle_handlers';
+
+export const registerCoreHandlers = (
+  registrar: LifecycleRegistrar,
+  config: HttpConfig,
+  env: Env
+) => {
+  jest.mock('./lifecycle_handlers', () => {
+    const actual = jest.requireActual('./lifecycle_handlers');
+    return {
+      ...actual,
+      createVersionCheckPostAuthHandler: jest.fn(actual.createVersionCheckPostAuthHandler),
+    };
+  });
+  jest.mock('./lifecycle_handlers', () => {
+    const actual = jest.requireActual('./lifecycle_handlers');
+    return {
+      ...actual,
+      createVersionCheckPostAuthHandler: jest.fn(actual.createVersionCheckPostAuthHandler),
+    };
+  });
+  jest.mock('./lifecycle_handlers', () => {
+    const actual = jest.requireActual('./lifecycle_handlers');
+    return {
+      ...actual,
+      createVersionCheckPostAuthHandler: jest.fn(actual.createVersionCheckPostAuthHandler),
+    };
+  });
+  jest.mock('./lifecycle_handlers', () => {
+    const actual = jest.requireActual('./lifecycle_handlers');
+    return {
+      ...actual,
+      createVersionCheckPostAuthHandler: jest.fn(actual.createVersionCheckPostAuthHandler),
+    };
+  });
+  jest.mock('./lifecycle_handlers', () => {
+    const actual = jest.requireActual('./lifecycle_handlers');
+    return {
+      ...actual,
+      createVersionCheckPostAuthHandler: jest.fn(actual.createVersionCheckPostAuthHandler),
+    };
+  });
+  jest.mock('./lifecycle_handlers', () => {
+    const actual = jest.requireActual('./lifecycle_handlers');
+    return {
+      ...actual,
+      createVersionCheckPostAuthHandler: jest.fn(actual.createVersionCheckPostAuthHandler),
+    };
+  });
+  // add headers based on config
+  registrar.registerOnPreResponse(createCustomHeadersPreResponseHandler(config));
+  // add extra request checks stuff
+  registrar.registerOnPostAuth(createXsrfPostAuthHandler(config));
+  if (config.versioned.strictClientVersionCheck !== false) {
+    // add check on version
+    registrar.registerOnPostAuth(createVersionCheckPostAuthHandler(env.packageInfo.version));
+  }
```

Comment:
nit: I wonder if we still want to warn via some custom header in the response.

This would help when troubleshooting HARs, or even, potentially using that on the browser to warn the user that they should refresh the page.

WDYT?

Reply:
I like that idea! As you suggest, this is one way to drive the "update available" UX we've spoken about... Currently it will only detect a new version when we ship a new stack version --- this will not be sufficient for more continuous updates. I think updating the Kibana version to contain a git hash or build nr or something would be needed...

I'd like to spend a bit more time thinking about how we want to approach that specifically. Do you think it is still worth adding a header like this now? If so, do you think we should just reuse the existing `kbn-version` header and set it to `update-available` or `out-of-sync`?

Reply:
Opened https://github.com/elastic/kibana/issues/159127

Reply:
IMO, we should just include the version in the header all the time, and add metadata to our logging-- not just error logging-- with the client version and server version.  That metadata will be invaluable to not only knowing if a mismatch is contributing to an error, but also if an error has spiked on a particular version of client + server, _and_ how many clients are "lagging" the upgrade curve.